File: experimental/adaptive-card-prompt-sample/dotnet/Dialogs/AdaptiveCardDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Linq;
using System;

namespace Microsoft.BotBuilderSamples
{
    public class AdaptiveCardDialog : ComponentDialog
    {

        public AdaptiveCardDialog()
            : base(nameof(AdaptiveCardDialog))
        {
            var adaptiveCardPrompt = new AdaptiveCardPrompt(nameof(AdaptiveCardPrompt));

            AddDialog(adaptiveCardPrompt);

            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
            {
                ShowCardAsync,
                DisplayResultsAsync
            }));

            InitialDialogId = nameof(WaterfallDialog);
        }

        private static async Task<DialogTurnResult> ShowCardAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var cardPath = Path.Combine("../Resources/", "adaptiveCard.json");
            var cardJson = File.ReadAllText(cardPath);
            var cardAttachment = new Attachment()
            {
                ContentType = "application/vnd.microsoft.card.adaptive",
                Content = JsonConvert.DeserializeObject(cardJson),
            };

            var promptSettings = new AdaptiveCardPromptSettings()
            {
                Card = cardAttachment
            };

            // Call the prompt
            return await stepContext.PromptAsync(nameof(AdaptiveCardPrompt), new PromptOptions()
            {
                Prompt = new Activity { Attachments = new List<Attachment>() { cardAttachment } }
            });
            
        }

        private async Task<DialogTurnResult> DisplayResultsAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // Use the result
            var result = stepContext.Result as JObject;
            var resultArray = result.Properties().Select(p => $"Key: {p.Name}  |  Value: {p.Value}").ToList();
            var resultString = string.Join("\n", resultArray);

            return await stepContext.EndDialogAsync();
        }
    }
}

File: experimental/adaptive-card-prompt-sample/javascript_nodejs/dialogs/adaptiveCardDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { AdaptiveCardPrompt } = require('../adaptiveCardPrompt');
const { ComponentDialog, DialogSet, DialogTurnStatus, WaterfallDialog } = require('botbuilder-dialogs');
const { CardFactory } = require('botbuilder');

const ADAPTIVE_CARD_DIALOG = 'ADAPTIVE_CARD_DIALOG';
const USER_PROFILE_PROPERTY = 'USER_PROFILE_PROPERTY';

class AdaptiveCardDialog extends ComponentDialog {
    constructor() {
        super(ADAPTIVE_CARD_DIALOG);

        // Load an adaptive card
        const cardJson = require('../resources/adaptiveCard.json');
        const card = CardFactory.adaptiveCard(cardJson);

        // Configure settings - All optional
        const promptSettings = { card: card };

        // Initialize the prompt
        const adaptiveCardPrompt = new AdaptiveCardPrompt('adaptiveCardPrompt', null, promptSettings);

        this.addDialog(adaptiveCardPrompt);

        this.addDialog(new WaterfallDialog(ADAPTIVE_CARD_DIALOG, [
            this.initialStep.bind(this),
            this.finalStep.bind(this)
        ]));

        this.initialDialogId = ADAPTIVE_CARD_DIALOG;
    }

    async initialStep(stepContext) {
        // Call the prompt
        return await stepContext.prompt('adaptiveCardPrompt');
    }

    async finalStep(stepContext) {
        // Use the result
        const result = stepContext.result;
        const resultString = Object.keys(result).map((key) => `Key: ${ key }   |   Value: ${ result[key] }`).join('\n');
        await stepContext.context.sendActivity(`Your input:\n\n${ resultString }`);
        return await stepContext.endDialog('You have finished the Adaptive Card Dialog. Happy coding!');
    }
}

module.exports.AdaptiveCardDialog = AdaptiveCardDialog;
module.exports.ADAPTIVE_CARD_DIALOG = ADAPTIVE_CARD_DIALOG;
